sdk/src/main/java/com/mapbox/maps/viewannotation/ViewAnnotationManagerImpl.kt

```diff
@@ -0,0 +1,250 @@
+    idsToAddSet.clear()
+    idsToDeleteSet.clear()
+    annotationMap.clear()
+    idLookupMap.clear()
+  }
+
+  private fun validateOptions(options: ViewAnnotationOptions) {
+    checkNotNull(options.geometry) { "Geometry can not be null!" }
+  }
+
+  private fun prepareViewAnnotation(inflatedView: View, options: ViewAnnotationOptions): View {
+    val inflatedViewLayout = inflatedView.layoutParams as FrameLayout.LayoutParams
+    val updatedOptions = options.toBuilder()
+      .width(if (options.width == null) inflatedViewLayout.width else options.width)
+      .height(if (options.height == null) inflatedViewLayout.height else options.height)
+      .build()
+    val viewAnnotation = ViewAnnotation(
+      view = inflatedView,
+      handleVisibility = options.visible == null,
+      viewLayoutParams = inflatedViewLayout,
+    )
+    val globalLayoutListener = ViewTreeObserver.OnGlobalLayoutListener {
+      if (viewAnnotation.handleVisibility) {
+        val isVisibleNow = inflatedView.visibility == View.VISIBLE
+        if (mapboxMap.getViewAnnotationOptions(viewAnnotation.id).value?.visible != isVisibleNow) {
+          mapboxMap.updateViewAnnotation(
+            viewAnnotation.id,
+            ViewAnnotationOptions.Builder()
+              .visible(isVisibleNow)
+              .build()
+          )
+        }
+      }
+    }
+    viewAnnotation.globalLayoutListener = globalLayoutListener
+    inflatedView.viewTreeObserver.addOnGlobalLayoutListener(globalLayoutListener)
+    annotationMap[viewAnnotation.id] = viewAnnotation
+    idLookupMap[inflatedView] = viewAnnotation.id
+    mapboxMap.addViewAnnotation(viewAnnotation.id, updatedOptions)
+    return inflatedView
+  }
+
+  private fun findByFeatureId(featureId: String): Pair<View?, ViewAnnotationOptions?> {
+    annotationMap.forEach { (id, annotation) ->
+      mapboxMap.getViewAnnotationOptions(id).value?.let { options ->
+        if (options.associatedFeatureId == featureId) {
+          return annotation.view to options
+        }
+      }
+    }
+    return null to null
+  }
+
+  private fun drawAnnotationViews(
+    positionDescriptorList: List<ViewAnnotationPositionDescriptor>
+  ) {
+    positionDescriptorMap.clear()
+    idsToRepositionSet.clear()
+    idsToAddSet.clear()
+    idsToDeleteSet.clear()
+
+    positionDescriptorList.forEachIndexed { i, descriptor ->
+      positionDescriptorMap[descriptor.identifier] =
+        ViewPosition(
+          leftTopCoordinate = descriptor.leftTopCoordinate,
+          width = descriptor.width,
+          height = descriptor.height,
+          zIndex = ViewPlugin.VIEW_PLUGIN_Z_TRANSLATION - 1f + i.toFloat() / positionDescriptorList.size
+        )
+    }
+    idsToRepositionSet.addAll(positionDescriptorMap.keys.intersect(currentViewsDrawnMap.keys))
+    idsToAddSet.addAll(positionDescriptorMap.keys.minus(currentViewsDrawnMap.keys))
+    idsToDeleteSet.addAll(currentViewsDrawnMap.keys.minus(positionDescriptorMap.keys))
```

Comment:
If it's called 60 fps then this 3 calls `intersect`/`minus` will spawn new collections every time, guess need to rewrite

Reply:
nice hint, will fix it

Reply:
Also - do we need to map PositionDescriptors to other separate objects? Could just pass it one by one [here](https://github.com/mapbox/mapbox-maps-android/pull/594/files/bd77e7ada7ca3cfe69d7aef707836ef00c8b70e9#diff-734ba43e94cb758b50aaa216439b68bffe47a7568f40aebfe3c27dcc4d3aeccbR193)